### PR TITLE
Commsignia Log Offloading Support

### PR DESCRIPTION
**Pull request by Michael7371**
# PR Details
## Description

This PR adds support for processing log offloading from Commsignia devices. The primary change here is supporting the new line characters that are included in these log messages, as well as parsing off the WSMP message header in the payload.

## Related Issue

To process offloaded logs in both CDOT/WYDOT deployment. In preparation of the anticipated firmware release that supports log offloading.

## How Has This Been Tested?

Added new unit tests to verify log processing of old messages (without newline characters).
Tested the log offloading of all .gz example log files in the data directory. To run this test, run the ODE through docker, at the minimum run the zookeeper, kafka, ODE, AEM, and ADM services. Copy all of the .gz files from the jpo-ode/data folder into the jpo-ode/uploads/bsmlog folder. This folder is monitored by the ODE and the files are ingested into the appropriate topics. If the processing succeeds the files are forwarded to the jpo-ode/uploads/backup folder and if they fail it goes to the jpo-ode/uploads/failed directory.
After processing the logs, run the following kafkacat command in WSL:

``` bash
sudo apt update
sudo apt install kafkacat
kafkacat -b localhost:9092 -G consumer_group topic.OdeBsmJson topic.OdeMapJson topic.OdeTimJson topic.OdeDriverAlertJson -f '\nTopic: %t\nKey: %k\nTimestamp: %T\nValue: %s\n'
```

This command will display the current offset of each topic listed in the consumer group. Verify that there are good counts for all topics listed. Here is an example output after processing the log files:

``` bash
% Waiting for group rebalance
% Group consumer_group rebalanced (memberid rdkafka-f7e42a50-1ef4-4473-af04-bbdcc79b9fc1): assigned: topic.OdeBsmJson [0], topic.OdeDriverAlertJson [0], topic.OdeMapJson [0], topic.OdeTimJson [0]
% Reached end of topic topic.OdeBsmJson [0] at offset 1723
% Reached end of topic topic.OdeMapJson [0] at offset 206
% Reached end of topic topic.OdeTimJson [0] at offset 279
% Reached end of topic topic.OdeDriverAlertJson [0] at offset 50
```

If you want, keep the kafkacat command running and process the log files one at a time to verify the output message looks good.

- [ x ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ x ] I have added any new packages to the sonar-scanner.properties file
- [ x ] My change requires a change to the documentation.
- [ x ] I have updated the documentation accordingly.
- [ x ] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/bugfix/Pull_request_template/docs/contributing_guide.md) 
- [ x ] I have added tests to cover my changes.
- [ x ] All new and existing tests passed.

### CODE REVIEW

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 1]
```diff
@@ -146,38 +155,37 @@ private void publishList(XmlUtils xmlUtils, List<OdeData> dataList) throws JsonP
...
-					//Nothing: If Spat log file and IEEE1609Cert is not present, Skip the Ieee1609Dot2Data encoding					
-				}
-				else 
-				{
-					Asn1Encoding msgEncoding = new Asn1Encoding("root", "Ieee1609Dot2Data", EncodingRule.COER);
-					msgMetadata.addEncoding(msgEncoding);
+
+				if (isSpatRecord() && msgMetadata instanceof OdeSpatMetadata
+						&& !((OdeSpatMetadata) msgMetadata).getIsCertPresent()) {
+					// Nothing: If Spat log file and IEEE1609Cert is not present, Skip the
```
Comment (drewjj) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:177]: You didn't write this but this bugs me. An 'if' statement to run nothing. This should just be written to check the negated value of this conditional to not have an unused 'if' statement and then move the 'else' code into it.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java [hunk 2]
```diff
@@ -90,4 +105,35 @@ public void writeTo(OutputStream os) throws IOException {
...
 
+
+
+   public byte[] removeHeader(byte[] packet) {
+      String hexPacket = HexUtils.toHexString(packet);
+      String hexPacketParsed = "";
+      for (String key : msgStartFlags.keySet()) {
+         String startFlag = msgStartFlags.get(key);
+         int startIndex = hexPacket.toLowerCase().indexOf(startFlag);
+         if (hexPacketParsed == ""){
```
Comment (drewjj) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:116]: You could use a break after a startIndex has been found that isn't -1 and the hexPacketParsed has been assigned a new value. The max number of iterations is 3 though so performance gains would be negligible. Just something to think about when writing a loop like this.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java [hunk 3]
```diff
@@ -90,4 +105,35 @@ public void writeTo(OutputStream os) throws IOException {
...
+            logger.debug("Start index for: " + key + " is: " + startIndex);
+            if (startIndex == -1) {
+               logger.debug("Message does not have header for: " + key);
+            } else if (startIndex <= HEADER_SIZE) {
+               logger.debug("Message has supported header. startIndex: " + startIndex + " msgFlag: " + startFlag);
+               hexPacketParsed = hexPacket;
+            } else if (startIndex > HEADER_SIZE && startIndex < 35) {
```
Comment (drewjj) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:123]: The 35 is fairly arbitrary looking from a perspective of having no clue about packet headers. May be a good idea to have a comment here describing the logic here.

**Review by drewjj — APPROVED** [review]
Nice job. Looks good. Just a couple of comments but I will approve it because it works!

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 4]
```diff
@@ -186,4 +191,40 @@ private void publishList(XmlUtils xmlUtils, List<OdeData> dataList) throws JsonP
...
+			payloadJson = JsonUtils.toJSONObject(payload.getData().toJson());
+			String hexPacket = payloadJson.getString("bytes");
+
+			for (String key : msgStartFlags.keySet()) {
+				String startFlag = msgStartFlags.get(key);
+				int startIndex = hexPacket.toLowerCase().indexOf(startFlag);
+				logger.debug("Start index for " + key + "(" + startFlag + ")" + " is: " + startIndex);
+				if (startIndex <= 20 && startIndex != 0 && startIndex != -1) {
+					logger.debug("Message has supported Ieee1609Dot2Data header, adding encoding rule to Asn1DecoderInput XML");
+					header = "Ieee1609Dot2Data";
```
Comment (payneBrandon) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:207]: If we've found the header here, should we break out of the loop to return? Is there any reason to continue at that point?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java [hunk 5]
```diff
@@ -186,4 +191,40 @@ private void publishList(XmlUtils xmlUtils, List<OdeData> dataList) throws JsonP
...
+				logger.debug("Payload JSON: " + payloadJson);
+			}
+		} catch (JsonUtilsException e) {
+			logger.error("JsonUtilsException while checking message header. Stacktrace: " + e.toString());
+
+		}
+		return header;
+	}
+
+	public BufferedInputStream removeNewLine(BufferedInputStream bis) {
```
Comment (payneBrandon) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:218]: It looks like this method just removes the newline if it's the next character in the input stream. Should we name this method to be a bit more specific? My first assumption was the method would take in the bufferedInputStream and remove all newline characters on a single call, but that seems to be incorrect.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java [hunk 6]
```diff
@@ -20,18 +20,31 @@
...
 import us.dot.its.jpo.ode.util.CodecUtils;
 
 public class PayloadParser extends LogFileParser {
 
-   public static final int PAYLOAD_LENGTH_LENGTH = 2;
+   private static Logger logger = LoggerFactory.getLogger(PayloadParser.class);
+   private static HashMap<String, String> msgStartFlags = new HashMap<String, String>();
 
+   private static final int HEADER_SIZE = 20; 
+   public static final int PAYLOAD_LENGTH_LENGTH = 2;
```
Comment (payneBrandon) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:37]: Copy/pasta or do we need the "_length_length" ?

Comment (payneBrandon) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:37]: Ah I see it was already in here above...any idea why the name?

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisherTest.java [hunk 7]
```diff
@@ -227,7 +228,8 @@ public void testPublishDriverAlertLogFile() throws Exception {
             (byte) 0xa4, (byte) 0xa1, (byte) 0x5c, (byte) 0xce, // 1.1 longitude
             (byte) 0x67, (byte) 0x06, (byte) 0x00, (byte) 0x00, // 1.2 elevation
             (byte) 0x04, (byte) 0x00, // 1.3 speed
-            (byte) 0x09, (byte) 0x27, // 1.4 heading
+            (byte) 0x09, (byte) 0x27, // 1.4 heading         assertEquals(expected, data.toJson());
```
Comment (payneBrandon) [inline jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisherTest.java:231]: Is this commented assert needed?

**Review by payneBrandon — CHANGES_REQUESTED** [review]
Looking good, just a couple comments/questions

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java (hunk 4 above)
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisher.java:207]: Nope, good idea!

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java (hunk 6 above)
Comment (Michael7371) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/importer/parser/PayloadParser.java:37]: I have no idea, haha! I can change it though.

File: jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisherTest.java (hunk 7 above)
Comment (Michael7371) [reply jpo-ode-svcs/src/test/java/us/dot/its/jpo/ode/coder/stream/LogFileToAsn1CodecPublisherTest.java:231]: Nope, must just be a remnant from some testing.

**Review by payneBrandon — APPROVED** [review]
lgtm!

**Pull request merged**